lib/telemetry/telemetry.go

```diff
@@ -0,0 +1,90 @@
+package telemetry
+
+// PLSREVIEW(kit): Should this package be lib/telemetry, or should we define it at the top level alongside agent
+//  and lib?
+
+import (
+	"time"
+
+	"github.com/armon/go-metrics"
+)
+
+type Label struct {
+	Key   string
+	Value string
+}
+
+// PLSREVIEW(kit): I decided to directly wrap the API of go-metrics because it informs all of our call sites.
+//  However, this does not need to be set in stone. For example, I renamed Incr to Inc. This is a good opportunity to
+//  adjust the metrics API we use internally. Any suggestions for improvements would be welcome!
+type MetricsClient interface {
```

Comment:
I think the shape of this interface is roughly correct, but I think we should move it to where it is consumed. There's a convention in Go that say "accept interfaces, return structs". I've been trying to do that more and more over the years and I've found it works very well.

In this case I think that means this `lib/telemetry` package would not define the interface. `Init` would return a `*telemetry.DefaultClient`, and consumers would define their own interface for the parts of that client that they need.  So for now this interface could probably be in `agent/setup.go` next to `BaseDeps` where it is used.

Re: naming, it might be easier to keep the names from go-metrics initially (ex: keep Incr instead of Inc). That way changing all the call sites is a little less work. Once everything is using the new type it should be a simple `Refactor -> Rename` in an IDE to change everything to the new name.

Reply:
I did some reading on the "accept interfaces, return structs" proverb to understand it a bit better. Coming from java, using interfaces is much stricter, so the patterns and expectations around their use is quite different. I believe this is a case where we _do_ in fact want to return an interface for the MetricsClient.

One of the goals of this is to abstract out the dependency on go-metrics so we have the ability to use other backends via the MetricsClient interface. For example, we may want to use opentelemetry, the other go-metrics lib that's based on CodeHale's metrics, or any client for specific metrics services e.g. prometheus.

The reason why we'd want to return an interface here, is because it meets the condition of our function needing to return a different type based on runtime conditions. (I sadly can't link to the section directly, but check the "Don't Return Concrete Types" section in this article, which goes over examples of when it's appropriate  https://www.integralist.co.uk/posts/go-interfaces/). In our case, the runtime condition is the config we receive during agent startup or reload and the type is type of client we're exposing - whether it's go-metrics (ArmonMetrics) or another.

Abstracting away that dependency should allow us to stop referencing go-metrics and prometheus within the Consul agent. I don't see a way to do that if we're implementing the client within the agent. Do you think that's the right goal to have for this PR, and if so, is this a good way to solve it?

Reply:
I reached out to Matt for opinions on this as well, and he agrees that we ought to define and implement the interface in the agent. I've been working on understanding this pattern better, and I think it's just a quality of Go that I need to get used to, even though it doesn't feel right to me yet!